providers/freebsd/host_freebsd_cgo.go

```diff
@@ -0,0 +1,263 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//go:build freebsd && cgo
+
+package freebsd
+
+import (
+	"context"
+	"errors"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/prometheus/procfs"
+
+	"github.com/elastic/go-sysinfo/internal/registry"
+	"github.com/elastic/go-sysinfo/providers/shared"
+	"github.com/elastic/go-sysinfo/types"
+)
+
+func init() {
+	registry.Register(newFreeBSDSystem())
+}
+
+type freebsdSystem struct{}
+
+func newFreeBSDSystem() freebsdSystem {
+	return freebsdSystem{}
+}
+
+func (s freebsdSystem) Host() (types.Host, error) {
+	return newHost()
+}
+
+type host struct {
+	procFS procFS
+	info   types.HostInfo
+}
+
+func (h *host) Info() types.HostInfo {
+	return h.info
+}
+
+func (h *host) CPUTime() (types.CPUTimes, error) {
+	cpu := types.CPUTimes{}
+	r := &reader{}
+	r.cpuTime(&cpu)
+	return cpu, r.Err()
+}
+
+func (h *host) Memory() (*types.HostMemoryInfo, error) {
+	m := &types.HostMemoryInfo{}
+	r := &reader{}
+	r.memInfo(m)
+	return m, r.Err()
+}
+
+func (h *host) FQDNWithContext(ctx context.Context) (string, error) {
+	return shared.FQDNWithContext(ctx)
+}
+
+func (h *host) FQDN() (string, error) {
+	return h.FQDNWithContext(context.Background())
+}
+
+func newHost() (*host, error) {
+	h := &host{}
+	r := &reader{}
+	r.architecture(h)
+	r.bootTime(h)
+	r.hostname(h)
+	r.network(h)
+	r.kernelVersion(h)
+	r.os(h)
+	r.time(h)
+	r.uniqueID(h)
+	return h, r.Err()
+}
+
+type reader struct {
+	errs []error
+}
+
+func (r *reader) addErr(err error) bool {
+	if err != nil {
+		if !errors.Is(err, types.ErrNotImplemented) {
+			r.errs = append(r.errs, err)
+		}
+		return true
+	}
+	return false
+}
+
+func (r *reader) Err() error {
+	if len(r.errs) > 0 {
+		return errors.Join(r.errs...)
+	}
+	return nil
+}
+
+func (r *reader) cpuTime(cpu *types.CPUTimes) {
+	times, err := cpuStateTimes()
+	if r.addErr(err) {
+		return
+	}
+	*cpu = *times
+}
+
+func (r *reader) memInfo(m *types.HostMemoryInfo) {
+	// Memory counter calculations:
+	//   total = physical memory
+	//   used = active + wired
+	//   free = free
+	//   available = buffers + inactive + cache + free
+
+	ps, err := pageSizeBytes()
+	if r.addErr(err) {
+		return
+	}
+	pageSize := uint64(ps)
```

Comment:
Can we not be optimistic here? At the moment, any single non-implemented collection means all subsequent collections are not performed. Is this guaranteed to be the behaviour we want? The alternative is to use the approach that is used for collections used in the linux `newHost` func where all are optimistically collected and then the aggregated errors are returned along with all the data that was obtained.

Reply:
I can either move the helpers onto `*reader` (pointer receiver) or we can also do this if permitted:

```go
func totalPhysicalMem(r *reader) uint64 {
	const mib = "hw.physmem"
	v, err := unix.SysctlUint64(mib)
	if r.addErr(err) {
		return 0
	}
	return v
}
```

Reply:
That looks OK.